File: backend/app/schemas/ai_observation.py
from datetime import datetime

from pydantic import BaseModel, Field


class AIProviderMetric(BaseModel):
    provider: str

    attempts: int

    successes: int

    errors: int

    success_rate: float = Field(ge=0, le=1)

    average_latency_ms: int

    p95_latency_ms: int


class AIFeatureMetric(BaseModel):
    feature: str

    total: int

    ai_successes: int

    rule_fallbacks: int

    errors: int

    success_rate: float = Field(ge=0, le=1)


class AIRecentEvent(BaseModel):
    feature: str

    provider: str

    model: str

    prompt_version: str

    status: str

    latency_ms: int

    error_category: str | None

    created_at: datetime


class AIMetrics(BaseModel):
    period_days: int

    generated_at: datetime

    total_feature_calls: int

    ai_successes: int

    rule_fallbacks: int

    errors: int

    success_rate: float = Field(ge=0, le=1)

    fallback_rate: float = Field(ge=0, le=1)

    provider_attempts: int

    prompt_versions: list[str]

    by_provider: list[AIProviderMetric]

    by_feature: list[AIFeatureMetric]

    recent_events: list[AIRecentEvent]

    privacy_notice: str

File: frontend/src/types/aiObservation.ts
export type AIProviderMetric = {
  provider: string;
  attempts: number;
  successes: number;
  errors: number;
  success_rate: number;
  average_latency_ms: number;
  p95_latency_ms: number;
};
export type AIFeatureMetric = {
  feature: string;
  total: number;
  ai_successes: number;
  rule_fallbacks: number;
  errors: number;
  success_rate: number;
};
export type AIRecentEvent = {
  feature: string;
  provider: string;
  model: string;
  prompt_version: string;
  status: string;
  latency_ms: number;
  error_category: string | null;
  created_at: string;
};
export type AIMetrics = {
  period_days: number;
  generated_at: string;
  total_feature_calls: number;
  ai_successes: number;

  rule_fallbacks: number;
  errors: number;
  success_rate: number;
  fallback_rate: number;

  provider_attempts: number;
  prompt_versions: string[];
  by_provider: AIProviderMetric[];

  by_feature: AIFeatureMetric[];
  recent_events: AIRecentEvent[];
  privacy_notice: string;
};
